cmake/modules/extensions.cmake

```diff
@@ -1552,40 +1552,62 @@ endfunction()
 #                     CMake namespace.
 function(import_kconfig prefix kconfig_fragment)
   cmake_parse_arguments(IMPORT_KCONFIG "" "TARGET" "" ${ARGN})
-  # Parse the lines prefixed with 'prefix' in ${kconfig_fragment}
   file(
     STRINGS
     ${kconfig_fragment}
     DOT_CONFIG_LIST
-    REGEX "^${prefix}"
     ENCODING "UTF-8"
   )
 
-  foreach (CONFIG ${DOT_CONFIG_LIST})
-    # CONFIG could look like: CONFIG_NET_BUF=y
-
-    # Match the first part, the variable name
-    string(REGEX MATCH "[^=]+" CONF_VARIABLE_NAME ${CONFIG})
+  foreach (LINE ${DOT_CONFIG_LIST})
+    if("${LINE}" MATCHES "^(${prefix}[^=]+)=([ymn]|.+$)")
+      # Matched a normal value assignment, like: CONFIG_NET_BUF=y
+      # Note: if the value starts with 'y', 'm', or 'n', then we assume it's a
+      # bool or tristate (we don't know the type from <kconfig_fragment> alone)
+      # and we only match the first character. This is to align with Kconfiglib.
+      set(CONF_VARIABLE_NAME "${CMAKE_MATCH_1}")
+      set(CONF_VARIABLE_VALUE "${CMAKE_MATCH_2}")
+    elseif("${LINE}" MATCHES "^# (${prefix}[^ ]+) is not set")
+      # Matched something like: # CONFIG_FOO is not set
+      # This is interpreted as: CONFIG_FOO=n
+      set(CONF_VARIABLE_NAME "${CMAKE_MATCH_1}")
+      set(CONF_VARIABLE_VALUE "n")
+    else()
+      # Ignore this line.
+      # Note: we also ignore assignments which don't have the desired <prefix>.
+      continue()
+    endif()
 
-    # Match the second part, variable value
-    string(REGEX MATCH "=(.+$)" CONF_VARIABLE_VALUE ${CONFIG})
-    # The variable name match we just did included the '=' symbol. To just get the
-    # part on the RHS we use match group 1
-    set(CONF_VARIABLE_VALUE ${CMAKE_MATCH_1})
+    # If the provided value is n, then the corresponding CMake variable or
+    # target property will be unset.
+    if("${CONF_VARIABLE_VALUE}" STREQUAL "n")
+      if(DEFINED IMPORT_KCONFIG_TARGET)
+        set_property(TARGET ${IMPORT_KCONFIG_TARGET} PROPERTY "${CONF_VARIABLE_NAME}")
+      else()
+        unset("${CONF_VARIABLE_NAME}" PARENT_SCOPE)
+      endif()
```

Comment:
I see that this part could benefit from config settings being internally represented by `n` instead of an `undef`.

I think having a prepare PR which changes the current behavior from `# CONFIG_<var> is not set` being undefined into being a `n` would make sense, and doing it as a prepare PR will allow us to run CI and check code and side-effect independent of other changes in this PR.

Reply:
> I see that this part could benefit from config settings being internally represented by `n` instead of an `undef`.

Sure. The way I wrote it, you could eventually remove this entire if-block and get the `n` behavior that you want.